### PR TITLE
Blocks are drawn on a second canvas on sleep

### PR DESCRIPTION
**Pull request by theroyalblah**
This should allow a much larger amount of blocks to be on screen at the same time by swapping out actual blocks with a static image only rendered a single time that overlays the first canvas. So when a block goes to sleep, it first becomes static, invisible, and finally is added to the second canvas.

In order for this to occur when the tab is loaded up again, when componentDidMount is called we have to iterate over the bodies and ensure they are re-added to the second canvas.

### CODE REVIEW

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 1]
```diff
@@ -43,6 +45,10 @@ class Scene extends React.Component {
       const loadedWorld = this.res.resurrect(world);
       Engine.merge(this.engine, { world: loadedWorld });
     }
+    const allBodies = Composite.allBodies(this.engine.world);
```
Comment (chewnoill) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: probably don't need to keep track of allBodies here, can just filter/map from here

File: app/javascript/components/widgets/numbers/sleeping-blocks.jsx [hunk 2]
```diff
@@ -55,13 +55,18 @@ Sleeping.update = function freezeOnSleep(bodies, timeScale) {
       ) {
         body.sleepCounter += 1;
 
-        if (body.sleepCounter >= body.sleepThreshold) {
+        if (body.sleepCounter >= body.sleepThreshold && body.render.visible) {
           // Bodies here set to static instead of being put to sleep
           Body.setStatic(body, true);
+          body.render.visible = false;
+          addToOverlay(body);
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/sleeping-blocks.jsx]: This function is doing a little more than just `freezeOnSleep` now. Its also hiding the bodies and calling a callback. I think function composition may be a useful pattern here.

File: app/javascript/components/widgets/numbers/sleeping-blocks.jsx [hunk 3]
```diff
@@ -55,13 +55,18 @@ Sleeping.update = function freezeOnSleep(bodies, timeScale) {
...
       } else if (body.sleepCounter > 0) {
         body.sleepCounter -= 1;
       }
     }
   }
-};
+}
+
+export default addToOverlay => (bodies, timeScale) =>
+  freezeOnSleep(bodies, timeScale, addToOverlay);
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/sleeping-blocks.jsx]: `curry` and `compose` may simplify this

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 4]
```diff
@@ -204,14 +230,26 @@ class Scene extends React.Component {
...
     return (
-      <div
-        style={{
-          position: 'fixed',
-          top: 0,
-          left: 0,
-        }}
-        ref={this.scene}
-      />
+      <div>
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: Prefer React.fragment to avoid unnecessary div

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 5]
```diff
@@ -143,6 +148,27 @@ class Scene extends React.Component {
     'count',
   ]);
 
+  addToOverlay = body => {
+    if (body && body.render.sprite.texture) {
+      const ctx = this.overlay.current.getContext('2d');
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Optimizing_canvas#Turn_off_transparency

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 6]
```diff
@@ -204,14 +230,26 @@ class Scene extends React.Component {
...
+        />
+        <canvas
+          width={window.innerWidth}
+          height={window.innerHeight}
+          style={{
+            position: 'fixed',
+            top: 0,
+            left: 0,
+          }}
+          ref={this.overlay}
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: Nit: this is technically an underlay.

File: app/javascript/components/widgets/numbers/sleeping-blocks.jsx [hunk 7]
```diff
@@ -55,13 +55,18 @@ Sleeping.update = function freezeOnSleep(bodies, timeScale) {
       ) {
         body.sleepCounter += 1;
 
-        if (body.sleepCounter >= body.sleepThreshold) {
+        if (body.sleepCounter >= body.sleepThreshold && body.render.visible) {
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/sleeping-blocks.jsx]: Why is `body.render.visible` necessary here? Is `sleepCounter` lost during serialization?

**Comment by cbarber** [comment]
I read this in a little more detail and I think the issue is the freezeOnSleep is doing a little too much since its currently a trivial way to extend the matterjs engine. It may be worthwhile to investigate the events emitted from the engine: `Events.trigger(engine, 'afterUpdate', event);` and `Events.trigger(engine, 'beforeUpdate', event);` to observe for a series of composed behaviors:
```
Event.observe(engine, ...., compose(
    setBodiesStaticIfSleeping,
    hideStaticBodies,
    copyStaticBodiesToUnderlay,
))
```

https://github.com/liabru/matter-js/blob/2ec247b7af1c6b5da6ee05c73274ed5822c73503/src/core/Engine.js#L230

File: app/javascript/components/widgets/numbers/falling-blocks.jsx (hunk 5 above)
Comment (theroyalblah) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: Doing this seems to destroy everything on the first canvas. Note the lack of counts at the top, and the blocks that are falling are not rendered either
Screen Shot 2019-06-24 at 2 55 50 PM

File: app/javascript/components/widgets/numbers/sleeping-blocks.jsx (hunk 7 above)
Comment (theroyalblah) [reply app/javascript/components/widgets/numbers/sleeping-blocks.jsx]: The objects that are currently "sleeping" pass the check, as they still exist but are simply invisible and static. Given that they pass this check, they would be redrawn to the second canvas as well as making unnecessary calculations. Technically this could be checking for if the body is static instead, as the body being invisible implies the body is static in our usage.

**Comment by theroyalblah** [comment]
@cbarber I tried implementing this, but splitting the functions into 3 and then composing them really damaged performance. What do you think about leaving sleeping-blocks and `freezeBlocks()` alone and creating a single member function that is used as follows
`Events.on(this.engine, 'beforeUpdate', b => this.prepareOverlay(b));`
This member function would make static blocks invisible and pass on singular blocks to ` addToOverlay()`. I have a working version of this and it seems to be rather performant and I think it solves the issue of freezeBlocks doing too much. Thoughts?

**Comment by cbarber** [comment]
Yeah, that sounds like an improvement. I'd like to dig into the performance issues with you for the composed functions. Could you grab the commit from the reflog and push it out on a separate branch?

File: app/javascript/components/widgets/numbers/falling-blocks.jsx (hunk 6 above)
Comment (theroyalblah) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: I think it is an overlay, as setting the background of the fallingBlocks to opaque canvas does not destroy this canvas, but the reverse is true. If the overlay's background is opaque, fallingBlocks is not visible. Unless I'm thinking about this backwards

Comment (cbarber) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: Oh derp. You're right. I got it backwards.

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 8]
```diff
@@ -23,6 +23,8 @@ class Scene extends React.Component {
     mutation: PropTypes.func.isRequired,
   };
 
+  imageCache = {};
```
Comment (chewnoill) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: Since its a separate idea, can we pull image caching into its own commit?

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 9]
```diff
@@ -177,6 +204,47 @@ class Scene extends React.Component {
...
+    const allBodies = Composite.allBodies(this.engine.world);
+    for (let i = 0; i < allBodies.length; i += 1) {
+      if (allBodies[i].isStatic) {
+        this.invisify(allBodies[i]);
+        this.addToOverlay(allBodies[i]);
+      }
+    }
+  }
+
+  // eslint-disable-next-line class-methods-use-this
```
Comment (chewnoill) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: what is this for?

Comment (theroyalblah) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: I wanted `invisify()` to be class member function but it doesn't need to be because it never uses `this`. So if you think it shouldn't be a member function I can change it.

Comment (cbarber) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: You can just move it to the class. There isn't any value for a funciton that does nothing with the class to be namespaced within the class.

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 10]
```diff
@@ -43,10 +46,20 @@ class Scene extends React.Component {
...
+      .map(b => this.addToOverlay(b));
+    Events.on(this.engine, 'beforeUpdate', b => this.prepareOverlay(b));
     this.addBlocks();
   }
 
   componentWillUnmount() {
+    const allBodies = Composite.allBodies(this.engine.world);
+    for (let i = 0; i < allBodies.length; i += 1) {
+      if (allBodies[i].onOverlay) {
+        allBodies[i].onOverlay = false;
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: This is a strange name for a boolean. `on<something>` is generally a callback. `is<something>` / `has<something>` are better boolean indicators

File: app/javascript/components/widgets/numbers/falling-blocks.jsx (hunk 8 above)
Comment (cbarber) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: Where is this used?

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 11]
```diff
@@ -177,6 +209,39 @@ class Scene extends React.Component {
...
+    if (!body || !body.render.sprite.texture || body.onOverlay) {
+      return;
+    }
+    const ctx = this.overlay.current.getContext('2d');
+    const { texture } = body.render.sprite;
+    const { sprite } = body.render;
+    ctx.translate(body.position.x, body.position.y);
+    ctx.rotate(body.angle);
+
+    const img = this.getImageFromCache(texture);
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: I'm not following the value of this. Module exports are already cached. What is the connection between `image` and `texture`?

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 12]
```diff
@@ -143,6 +160,33 @@ class Scene extends React.Component {
...
+      b.render.visible = false;
+    }
+  };
+
+  invisify = body => {
+    if (body.render.visible) {
+      const b = body;
+      b.render.visible = false;
+    }
+  };
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: This doesn't relate to the class so it can be extracted. 

Also can likely be simplified with a ramda method.

Also duplicated

File: app/javascript/components/widgets/numbers/falling-blocks.jsx (hunk 8 above)
Comment (chewnoill) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: https://github.com/mojotech/helios2/pull/128/commits/fc0a94a5df76082bb99c0bc477ffee01ea723713#diff-ed8e5fea165eae9a5f4092f20661d47fR164

File: app/javascript/components/widgets/numbers/falling-blocks.jsx (hunk 11 above)
Comment (theroyalblah) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: It's to try and avoid `image = new Image();` which seemed to be very expensive. Texture is pulled from the body object and contains the filepath for the image.

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 13]
```diff
@@ -43,10 +51,23 @@ class Scene extends React.Component {
       const loadedWorld = this.res.resurrect(world);
       Engine.merge(this.engine, { world: loadedWorld });
     }
+    Composite.allBodies(this.engine.world)
+      .filter(body => body.isStatic && body.render.sprite.texture)
+      .map(b => this.addToOverlay(b));
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: This should be a `forEach`, not `map`

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 14]
```diff
@@ -43,10 +51,23 @@ class Scene extends React.Component {
...
     this.addBlocks();
   }
 
   componentWillUnmount() {
+    const allBodies = Composite.allBodies(this.engine.world);
+    allBodies
+      .filter(body => body.onOverlay)
+      .map(body => {
+        const b = body;
+        b.onOverlay = false;
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: `is<something>`, `has<something>`, `should<something>` are better hints at bools. `on<something` suggests a callback.

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 15]
```diff
@@ -159,24 +195,61 @@ class Scene extends React.Component {
...
+      img,
+      img.width * -sprite.xOffset,
+      img.height * -sprite.yOffset,
+    );
+
+    ctx.rotate(-body.angle);
+    ctx.translate(-body.position.x, -body.position.y);
+    const b = body;
+    b.isOnOverlay = true;
+    Events.off(body, 'sleepStart', overlayBody =>
```
Comment (chewnoill) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: not sure Events.off gets a callback

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 16]
```diff
@@ -190,6 +263,15 @@ class Scene extends React.Component {
 
   serialise(object) {
     return this.res.stringify(object, (key, value) => {
+      // serialise fails if it attempts to work on a function, even if it's a member of an object
+      // to prevent this from occuring, we need to check objects for their first member to see if it's a function
+      if (
+        value &&
+        typeof value === typeof object &&
+        typeof value[0] === typeof (() => {})
```
Comment (chewnoill) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: can we check the `key` instead of the `typeof value[0]`?

File: app/javascript/components/widgets/numbers/falling-blocks.jsx (hunk 15 above)
Comment (theroyalblah) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: Looks like it:
    /**
     * Removes the given event callback. If no callback, clears all callbacks in `eventNames`. If no `eventNames`, clears all events.
     * @method off
     * @param {} object
     * @param {string} eventNames
     * @param {function} callback
     */
`    Events.off = function(object, eventNames, callback) {`

https://github.com/liabru/matter-js/blob/2ec247b7af1c6b5da6ee05c73274ed5822c73503/src/core/Events.js#L45

File: app/javascript/components/widgets/numbers/falling-blocks.jsx (hunk 16 above)
Comment (theroyalblah) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: This code also works, if you think it would be better:
`if (key === 'sleepStart') {
return null;
      }`

File: app/javascript/components/widgets/numbers/falling-blocks.jsx (hunk 15 above)
Comment (chewnoill) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: do we need to `addToOverlay` on this callback?

File: app/javascript/components/widgets/numbers/falling-blocks.jsx (hunk 16 above)
Comment (chewnoill) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: I would prefer to ignore all `key === 'events'` if possible

File: app/javascript/components/widgets/numbers/falling-blocks.jsx (hunk 15 above)
Comment (theroyalblah) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: nope, it's gone now

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 17]
```diff
@@ -143,6 +182,28 @@ class Scene extends React.Component {
     'count',
   ]);
 
+  getImageFromCache(imagePath) {
```
Comment (chewnoill) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: looks like this got smooshed into the wrong commit

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 18]
```diff
@@ -143,6 +182,15 @@ class Scene extends React.Component {
     'count',
   ]);
 
+  sleepStartEvent = block => {
+    Events.on(block, 'sleepStart', () => {
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/falling-blocks.jsx]: Javascript events typically bubble up so you may be able to listen on the `world` to reduce creating/destroying events for every body.

Comment (cbarber) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: Just realized this is internal to matterjs and not the javascript Event API so this isn't true.

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 19]
```diff
@@ -159,24 +207,56 @@ class Scene extends React.Component {
...
+    ctx.drawImage(
+      img,
+      img.width * -sprite.xOffset,
+      img.height * -sprite.yOffset,
+    );
+
+    ctx.rotate(-body.angle);
+    ctx.translate(-body.position.x, -body.position.y);
+    const b = body;
+    b.isOnOverlay = true;
```
[the remark this answers is not in the document]
Comment (theroyalblah) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: eslint doesn't like assigning to function parameters

> Assignment to property of function parameter 'body'.eslint(no-param-reassign)

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 20]
```diff
@@ -18,6 +26,22 @@ const wallWidth = 50;
 // Padding applied to the height of the walls to prevent blocks created contemporaneously from colliding.
 const padding = 50;
 
+// eslint-disable-next-line no-underscore-dangle
+Sleeping._motionSleepThreshold = 0.5;
```
[the remark this answers is not in the document]
Comment (chewnoill) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: looks like this was moved in from `@numbers/sleeping-blocks'

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 21]
```diff
@@ -43,10 +68,23 @@ class Scene extends React.Component {
       const loadedWorld = this.res.resurrect(world);
       Engine.merge(this.engine, { world: loadedWorld });
     }
+    Composite.allBodies(this.engine.world)
+      .filter(body => !body.isStatic)
+      .map(block => this.sleepStartEvent(block));
```
[the remark this answers is not in the document]
Comment (chewnoill) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: why is it unnecessary? or do you just prefer foreach?

File: app/javascript/components/widgets/numbers/falling-blocks.jsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (cbarber) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: I think in this case I'd prefer `// eslint-disable-next-line no-param-reassign` so we're not obscuring that we're mutating the object.

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 22]
```diff
@@ -190,6 +270,11 @@ class Scene extends React.Component {
 
   serialise(object) {
     return this.res.stringify(object, (key, value) => {
+      // serialise fails if it attempts to work on a function, even if it's a member of an object
+      // to prevent this from occuring, we need to check if there was an event
+      if (key === 'events') {
```
[the remark this answers is not in the document]
Comment (chewnoill) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: This implementation prevents serialize anything from the `events` branches of the object; although the comment suggests a more generalized intent.  I'm fine with either approach, the comment is a little confused though.

File: app/javascript/components/widgets/numbers/falling-blocks.jsx [hunk 23]
```diff
@@ -204,14 +289,26 @@ class Scene extends React.Component {
...
     return (
-      <div
-        style={{
-          position: 'fixed',
-          top: 0,
-          left: 0,
-        }}
-        ref={this.scene}
-      />
+      <React.Fragment>
```
[the remark this answers is not in the document]
Comment (chewnoill) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: agreed, if it works.

File: app/javascript/components/widgets/numbers/falling-blocks.jsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (theroyalblah) [reply app/javascript/components/widgets/numbers/falling-blocks.jsx]: it's a foreach now
